python/addons/embed_bert_pytorch.py

```diff
@@ -70,6 +82,263 @@ def run(self, tokens, vocab):
+        self.delim = kwargs.get('token_delim', '~~')
+
+    def iterable(self, tokens):
+        return super().iterable(_token_iterator(self, tokens))
+
+
+@register_vectorizer(name='wordpiece1d-with-heads-dict')
+class WordPieceVectorizer1DWithHeadsDict(WordPieceVectorizer1D):
+    """This vectorizer produces the tokens and the index of the heads at the same time.
+
+    In this one the results are output as a dict
+    """
+
+    def iterable(self, tokens):
+        yield '[CLS]', 0
+        for tok in tokens:
+            if tok == '<unk>':
+                yield '[UNK]', 1
+            elif tok == '<EOS>':
+                yield '[SEP]', 0
+            else:
+                for i, subtok in enumerate(self.tokenizer.tokenize(tok)):
+                    if i == 0:
+                        yield subtok, 1
+                    else:
+                        yield subtok, 0
+        yield '[SEP]', 0
+
+    def _next_element(self, tokens, vocab):
+        for atom, head in self.iterable(tokens):
+            value = vocab.get(atom, vocab['[UNK]'])
+            yield value, head
+
+    def run(self, tokens, vocab):
+        self.mxlen = self.max_seen if self.mxlen < 0 else self.mxlen
+        vec1d = np.zeros(self.mxlen, dtype=np.long)
+        heads = np.zeros(self.mxlen, dtype=np.long)
+        for i, (atom, head) in enumerate(self._next_element(tokens, vocab)):
+            if i == self.mxlen:
+                i -= 1
+                break
+            vec1d[i] = atom
+            heads[i] = head
+        heads = mask_to_index(heads)
+        valid_length = i + 1
+        return {'tokens': vec1d, 'heads': heads}, valid_length
+
+
+@register_vectorizer(name="dict-wordpiece1d-with-heads-dict")
+class DictWordPieceVectoizer1DWithHeadsDict(WordPieceVectorizer1DWithHeadsDict):
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.fields = listify(kwargs.get('fields', 'text'))
+        self.delim = kwargs.get('token_delim', '~~')
+
+    def iterable(self, tokens):
+        return super().iterable(_token_iterator(self, tokens))
+
+
+@register_vectorizer(name='heads-wordpiece1d')
+class HeadedWordPieceVectorizer1D(WordPieceVectorizer1D):
+    """Output a list of indices that are the heads of the subword tokens."""
+
+    def iterable(self, tokens):
+        j = 1  # Don't point at [CLS]
+        for tok in tokens:
+            # Unk is something we want to label
+            if tok == '<unk>':
+                yield j
+            # We don't want to label [SEP]
+            elif tok == '<EOS>':
+                pass
+            else:
+                for i, subtok in enumerate(self.tokenizer.tokenize(tok)):
+                    if i == 0:
+                        yield j  # Head
+                    else:
+                        j += 1
+            j += 1
+
+    def _next_element(self, tokens, vocab):
+        for atom in self.iterable(tokens):
+            yield atom
+
+    def run(self, tokens, vocab):
```

Comment:
This needs to be test for the case where mxlen is less than the length of the example. 

It might need to refactor so instead of the i == mxlen break check it might need to check if the head index is >= mxlen. Or else it might need to do a clamp so one of the indices isn't larger than the mxlen